src/main/kotlin/it/unibo/protelis/web/execution/simulated/SimulatedProtelisEngine.kt

```diff
@@ -0,0 +1,149 @@
+package it.unibo.protelis.web.execution.simulated
+
+import it.unibo.alchemist.core.implementations.Engine
+import it.unibo.alchemist.core.interfaces.Simulation
+import it.unibo.alchemist.core.interfaces.Status
+import it.unibo.alchemist.loader.YamlLoader
+import it.unibo.alchemist.model.implementations.positions.Euclidean2DPosition
+import it.unibo.alchemist.model.implementations.times.DoubleTime
+import it.unibo.alchemist.model.interfaces.Position
+import it.unibo.protelis.web.execution.CoroutineProtelisEngine
+import it.unibo.protelis.web.execution.ProtelisObserver
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.future.asCompletableFuture
+import kotlinx.coroutines.launch
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+class SimulatedProtelisEngine() : CoroutineProtelisEngine() {
+  private val logger: Logger = LoggerFactory.getLogger(this::class.java)
+  private var alchemistEngine: AtomicReference<Simulation<Any /* TODO */, Euclidean2DPosition>?> = AtomicReference(null)
+
+  init {
+    logger.debug("Simulation container created")
+  }
+
+  constructor(sourceCode: String, monitor: ProtelisObserver) : this() {
+    GlobalScope.launch { setupAwait(sourceCode, monitor) }.asCompletableFuture().join()
+  }
+
+  override suspend fun setupAwait(sourceCode: String, monitor: ProtelisObserver) {
+    if (alchemistEngine.get() != null) stopAwait()
+    val sim = setupSimulationAsync().await()
+    alchemistEngine.set(sim)
+    setCode(sourceCode)
+    alchemistEngine.get()
+      ?.addOutputMonitor(ProtelisUpdateAdapter(monitor))
+      ?: throw IllegalStateException("Simulation was not set up correctly")
+    alchemistEngine.get()
+      ?.awaitFor(Status.INIT)
+      ?: throw IllegalStateException("Simulation was not set up correctly")
+    GlobalScope.launch { sim.run() }
+    logger.debug("Simulation Engine set up correctly")
+  }
+
+  override suspend fun startAwait() {
+    checkNotNull(alchemistEngine.get())
+    alchemistEngine.get()?.play()
+    alchemistEngine.get()?.awaitFor(Status.RUNNING) // TODO: should I wait RUNNING instead?
+    logger.debug("Simulation started")
+  }
+
+  override suspend fun stopAwait() {
+    checkNotNull(alchemistEngine.get())
+    alchemistEngine.get()?.terminate()
+    alchemistEngine.get()?.awaitFor(Status.TERMINATED)
+    alchemistEngine.set(null)
+    logger.debug("Simulation stopped and deleted")
+  }
+
+  /** Asynchronously build a new simulation from a loader of YAML files. */
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [CommentOverPrivateFunction - [setupSimulationAsync] at /src/src/main/kotlin/it/unibo/protelis/web/execution/simulated/SimulatedProtelisEngine.kt:64:3](https://app.codacy.com/manual/NiccoMlt/Protelis-Web/pullRequest?prid=4969516)